src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -99,38 +99,45 @@ public long add(DecoratedKey key, Clustering clustering, ByteBuffer value)
         synchronized (writeLock)
         {
             AbstractAnalyzer analyzer = columnContext.getAnalyzer();
-            value = TypeUtil.encode(value, validator);
-            analyzer.reset(value.duplicate());
-            final PrimaryKey primaryKey = PrimaryKey.of(key, clustering);
-            final long initialSizeOnHeap = data.sizeOnHeap();
-            final long initialSizeOffHeap = data.sizeOffHeap();
-            final long reducerHeapSize = primaryKeysReducer.heapAllocations();
-
-
-            while (analyzer.hasNext())
+            try
             {
-                final ByteBuffer term = analyzer.next();
-                setMinMaxTerm(term);
-
-                final ByteComparable encodedTerm = encode(term);
-                try
+                value = TypeUtil.encode(value, validator);
+                analyzer.reset(value.duplicate());
+                final PrimaryKey primaryKey = PrimaryKey.of(key, clustering);
+                final long initialSizeOnHeap = data.sizeOnHeap();
+                final long initialSizeOffHeap = data.sizeOffHeap();
+                final long reducerHeapSize = primaryKeysReducer.heapAllocations();
+
+                while (analyzer.hasNext())
                 {
-                    if (term.limit() <= MAX_RECURSIVE_KEY_LENGTH)
+                    final ByteBuffer term = analyzer.next();
+                    setMinMaxTerm(term);
+
+                    final ByteComparable encodedTerm = encode(term);
+                    try
                     {
-                        data.putRecursive(encodedTerm, primaryKey, primaryKeysReducer);
+                        if (term.limit() <= MAX_RECURSIVE_KEY_LENGTH)
+                        {
+                            data.putRecursive(encodedTerm, primaryKey, primaryKeysReducer);
+                        }
+                        else
+                        {
+                            data.apply(Trie.singleton(encodedTerm, primaryKey), primaryKeysReducer);
+                        }
                     }
-                    else
+                    catch (MemtableTrie.SpaceExhaustedException e)
                     {
-                        data.apply(Trie.singleton(encodedTerm, primaryKey), primaryKeysReducer);
+                        //TODO Handle this properly
```

Comment:
This has been left over from another ticket but we do need to figure out what should be done at this point rather than just throwing.